client/worker_grpc.go

```diff
@@ -94,7 +94,7 @@ func (c *TaskHubGrpcClient) processActivityWorkItem(
 	executor backend.Executor,
 	req *protos.ActivityRequest,
 ) {
-	var tc *protos.TraceContext = nil // TODO: How to populate trace context?
+	var tc *protos.TraceContext = req.ParentTraceContext
```

Comment:
I need more help understanding why we need to pass `ParentTraceContext` on this request object instead of just getting it directly from the gRPC context. Is it because work items are received on a stream and gRPC doesn't automatically populate any distributed trace context?

Reply:
I think 
1. the grpc `context` cannot directly used to pass through key-value pair (span is usually stored as key-val pair in context in go), instead it needs a bit more work https://www.hward.com/golang-grpc-context-client-server/. 
2. we obtain the stream at https://github.com/microsoft/durabletask-go/blob/fc5756737d3c85ac4903adf35b58fce25f36641b/client/worker_grpc.go#L26, at this point we don't have any trace context set up in the context, the context basically is an empty context. The stream should not have any trace info therefore. 

I tried a local run to assert this
This is the context we build up on grpc server side
![image](https://github.com/microsoft/durabletask-go/assets/89094811/6e21ae8d-ab73-4f04-97c0-4f8e9662768e)

This is the context at https://github.com/microsoft/durabletask-go/blob/fc5756737d3c85ac4903adf35b58fce25f36641b/client/worker_grpc.go#L93
![image](https://github.com/microsoft/durabletask-go/assets/89094811/e96e8d8e-d08b-4e9a-8633-896489f40645)

Reply:
Copying from here: https://github.com/microsoft/durabletask-go/issues/31#issuecomment-1886219083

Agree @kaibocai. I looked into it as well. We can easily propagate context for a normal grpc request(non-streaming) using otel interceptors(e.g. [here](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/3cecdcf33c4701a3abd8557d38ffde090a23b696/instrumentation/google.golang.org/grpc/otelgrpc/example/client/main.go#L81)) but for streaming requests, context is propagated at the start only i.e. when the stream is [created](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/3cecdcf33c4701a3abd8557d38ffde090a23b696/instrumentation/google.golang.org/grpc/otelgrpc/example/client/main.go#L97). Once the stream creation is done, then since no context is passed during [stream.send()](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/3cecdcf33c4701a3abd8557d38ffde090a23b696/instrumentation/google.golang.org/grpc/otelgrpc/example/client/main.go#L103) method, and because we can't update stream context, the newer context(updated after stream creation) is never passed to client/server.

Here in DTF-Go as well at the time of [stream creation](https://github.com/microsoft/durabletask-go/blob/fc5756737d3c85ac4903adf35b58fce25f36641b/client/worker_grpc.go#L26), there is nothing in context(since no orchestration/acitivity are created yet). We have the updated context(with all span info) when activity/orchestration workitem [needs to be sent to the app](https://github.com/microsoft/durabletask-go/blob/fc5756737d3c85ac4903adf35b58fce25f36641b/backend/executor.go#L136C42-L136C42) but this context is not passed while [sending the workItem](https://github.com/microsoft/durabletask-go/blob/fc5756737d3c85ac4903adf35b58fce25f36641b/backend/executor.go#L218). So best way to do it is to encode trace context in the message itself.